gh-73069: fix race conditions in os.scandir and associated direntry

duaneg opened the pull request:
This fixes the first two (c1 & c2) crashes from gh-73069. Let me know if I should add a news entry.

duaneg committed:
gh-73069 c1: fix race conditions with os.scandir iterator

duaneg committed:
gh-73069 c2: fix race conditions with os.scandir direntry stat

duaneg committed:
NEWS entry for gh-73069

ZeroIntensity reviewed (commented):
- A blurb/news entry makes sense here.
- Mutexes are prone to re-entrancy deadlocks in cases where we call back into Python, *e.g.*, in an object's finalizer. We should be using [critical sections](https://docs.python.org/3/c-api/init.html#python-critical-section-api), probably alongside [argument clinic](https://devguide.python.org/development-tools/clinic/#how-to-use-critical-sections-with-argument-clinic).

duaneg committed:
Switch to _PyRecursiveMutex from PyMutex

It seems we do need to worry about re-entrancy, so switch to using a recursive
mutex. Note we can't use a critical section since we need to perform blocking
IO while holding the lock, which critical section would implicitly release.

duaneg committed:
Refactor so mutex is held for minimal time and not over any code that could be
re-entrant.

duaneg committed:
Refactor to use atomics & compare/exchange instead of holding a lock to avoid
possibility of deadlock upon re-entrance.

duaneg committed:
Refactor to share code and eliminate duplication between Windows & POSIX platforms

duaneg committed:
struct dirent is variable length in the original POSIX specification (the final
filename field being an unsized array) and in actual practice on the WASI
platform. Thus just copying one into a stack-allocated structure is not good
enough.

The obvious way to address this is to allocate memory on the heap and return
that. However, it seems wasteful to dynamically allocate a temporary chunk of
data just to copy some data in and then out again.

Instead, if the name field is potentially not long enough, allocate some extra
space *on the stack* following the structure such that it must be sufficient to
hold the the full record length (and copy all of that, not just the potentially
smaller fixed structure size).

duaneg committed:
Merge remote-tracking branch 'origin/main' into gh-73069

duaneg committed:
Add regression tests
